pkg/storage/fingerprint_writer.go

```diff
@@ -0,0 +1,141 @@
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/errors"
+)
+
+// fingerprintWriter hashes every key/timestamp and value for point keys, and
+// combines their hashes via a XOR into a running aggregate.
+//
+// Range keys are not fingerprinted but instead written to a pebble SST that is
+// returned to the caller. This is because range keys do not have a stable,
+// discrete identity and so it is up to the caller to define a deterministic
+// fingerprinting scheme across all returned range keys.
+//
+// The caller must Finish() and Close() the fingerprintWriter to finalize the
+// writes to the underlying pebble SST.
+type fingerprintWriter struct {
+	hasher hash.Hash64
+
+	sstWriter *SSTWriter
+	xorAgg    *uintXorAggregate
+}
+
+// makeFingerprintWriter creates a new fingerprintWriter.
+func makeFingerprintWriter(
+	ctx context.Context, hasher hash.Hash64, cs *cluster.Settings, f io.Writer,
+) fingerprintWriter {
+	// TODO(adityamaru,dt): Once
+	// https://github.com/cockroachdb/cockroach/issues/90450 has been addressed we
+	// should write to a kvBuf instead of a Backup SST writer.
+	sstWriter := MakeBackupSSTWriter(ctx, cs, f)
+	return fingerprintWriter{
+		sstWriter: &sstWriter,
+		hasher:    hasher,
+		xorAgg:    &uintXorAggregate{},
+	}
+}
+
+type uintXorAggregate struct {
+	sum uint64
+}
+
+// add inserts one value into the running xor.
+func (a *uintXorAggregate) add(x uint64) {
+	a.sum = a.sum ^ x
+}
+
+// result returns the xor.
+func (a *uintXorAggregate) result() uint64 {
+	return a.sum
+}
+
+// Finish finalizes the underlying SSTWriter, and returns the aggregated
+// fingerprint for point keys.
+func (f *fingerprintWriter) Finish() (uint64, error) {
+	if err := f.sstWriter.Finish(); err != nil {
+		return 0, err
+	}
+	return f.xorAgg.result(), nil
+}
+
+// Close finishes and frees memory and other resources. Close is idempotent.
+func (f *fingerprintWriter) Close() {
+	if f.sstWriter == nil {
+		return
+	}
+	f.sstWriter.Close()
+	f.hasher.Reset()
+	f.xorAgg = nil
+	f.sstWriter = nil
+}
+
+var _ ExportWriter = &fingerprintWriter{}
+
+// PutRawMVCCRangeKey implements the Writer interface.
+func (f *fingerprintWriter) PutRawMVCCRangeKey(key MVCCRangeKey, bytes []byte) error {
+	// We do not fingerprint range keys, instead, we write them to a Pebble SST.
+	// This is because range keys do not have a stable, discrete identity and so
+	// it is up to the caller to define a deterministic fingerprinting scheme
+	// across all returned range keys.ler to decide how to fingerprint them.
+	return f.sstWriter.PutRawMVCCRangeKey(key, bytes)
+}
+
+// PutRawMVCC implements the Writer interface.
+func (f *fingerprintWriter) PutRawMVCC(key MVCCKey, value []byte) error {
+	defer f.hasher.Reset()
+
+	// Hash the key/timestamp and value of the RawMVCC.
+	if err := f.hash(key.Key); err != nil {
+		return err
+	}
+	if err := f.hash([]byte(key.Timestamp.String())); err != nil {
```

Comment:
This doesn't seem great, since the string representation isn't guaranteed to be stable nor lossy. How about hashing the encoded MVCC timestamp instead? We can use e.g. `encodeMVCCTimestampToBuf()` to avoid allocations.

Reply:
makes sense, changed to encoding the buf instead. In the follow up that introduces the builtin that actually uses the fingerprint mode I was writing queries like:

```
WITH sfp AS (SELECT xor_agg(fnv64(key, ts::BYTES, value)) AS scanfp from crdb_internal.scan(crdb_internal.table_span('t'::regclass::int)))
  SELECT count(*) FROM crdb_internal.fingerprint(crdb_internal.table_span('t'::regclass::int),'$before_insert'::TIMESTAMPTZ, false) AS efp, sfp WHERE efp = sfp.scanfp;`
```

to validate that a scan and an exportrequest returned the same fingerprint. This won't work out of the box cause `crdb_internal.scan` returns the `String()` version of the timestamp. Maybe we can add an overload to `crdb_internal.scan` that returns the encoded MVCC timestamp instead.

Reply:
By the way, I thought `encodeMVCCTimestampToBuf` would resize the buffer, but I see that it doesn't because of how it's used in the key encoding functions. This code is very performance-sensitive, so I can't add in additional size checks in the hot path. But here's a patch which adds `EncodeMVCCTimestampToBuf` that does resize the buffer (the name duplication isn't great, but at least it's consistent with `EncodeMVCCKeyToBuf` / `encodeMVCCKeyToBuf`).

[encodets.patch.txt](https://github.com/cockroachdb/cockroach/files/9960058/encodets.patch.txt)

Reply:
thanks for the patch, added it in.